hazelcast/src/main/java/com/hazelcast/map/impl/LazyMapEntry.java

```diff
@@ -0,0 +1,170 @@
+        checkNotNull(key, "key cannot be null");
+        checkNotNull(serializationService, "SerializationService cannot be null");
+
+        keyData = null;
+        keyObject = null;
+
+        if (key instanceof Data) {
+            this.keyData = (Data) key;
+        } else {
+            this.keyObject = key;
+        }
+
+        valueData = null;
+        valueObject = null;
+
+        if (value instanceof Data) {
+            this.valueData = (Data) value;
+        } else {
+            this.valueObject = value;
+        }
+
+        this.serializationService = serializationService;
+    }
+
+    @Override
+    public Object getKey() {
+        if (keyObject == null && serializationService != null) {
+            keyObject = serializationService.toObject(keyData);
+        }
+        return keyObject;
+    }
+
+
+    @Override
+    public Object getValue() {
+        // It is required to check null serializationService because
+        // after subsequent serialization and deserialization of a LazyMapEntry instance
+        // serializationService will be null.
+        if (valueObject == null && serializationService != null) {
+            valueObject = serializationService.toObject(valueData);
+        }
+        return valueObject;
+    }
+
+    @Override
+    public Object setValue(Object value) {
+        modified = true;
+        Object oldValue = getValue();
+        this.valueObject = value;
+        this.valueData = null;
+        return oldValue;
+    }
+
+    public Object getKeyData() {
+        if (keyData == null && serializationService != null) {
+            keyData = serializationService.toData(keyObject);
+        }
+        return keyData;
+    }
+
+    public Object getValueData() {
+        if (valueData == null && serializationService != null) {
+            valueData = serializationService.toData(valueObject);
+        }
+        return valueData;
+    }
+
+    public void setSerializationService(SerializationService serializationService) {
+        this.serializationService = checkNotNull(serializationService, "SerializationService cannot be null");
+    }
+
+    public boolean isModified() {
+        return modified;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (!(o instanceof Map.Entry)) {
+            return false;
+        }
+        Map.Entry<?, ?> e = (Map.Entry<?, ?>) o;
+        return equals(getKey(), e.getKey()) && equals(getValue(), e.getValue());
+    }
+
+    private static boolean equals(Object o1, Object o2) {
+        return o1 == null ? o2 == null : o1.equals(o2);
+    }
+
+    @Override
+    public int hashCode() {
+        return (getKey() == null ? 0 : getKey().hashCode())
+                ^ (getValue() == null ? 0 : getValue().hashCode());
+    }
+
+    @Override
+    public String toString() {
+        return getKey() + "=" + getValue();
+    }
+
+    private void writeObject(java.io.ObjectOutputStream out)
```

Comment:
`EntryProcessor#process(entry)` may return `LazyMapEntry` instance as a response to user but there is a possibility that he did not trigger on-demand deserialization, means not called `LazyMapEntry#getKey` or `LazyMapEntry#getValue`, in that case, `LazyMapEntry#keyObject` will be null but actually it may not be null. Trying to prevent that anomaly by overriding them.